crates/core/src/network/connection_manager.rs

```diff
@@ -596,37 +603,40 @@ where
 
         // body are not allowed to be empty.
         let Some(body) = msg.body else {
-            connection
-                .send_control_frame(ConnectionControl::codec_error("Body is missing on message"));
+            tx.unbounded_drain(DrainReason::CodecError(
+                "Body is missing on message".to_owned(),
+            ));
             break;
         };
 
         // Welcome and hello are not allowed after handshake
         if body.is_welcome() || body.is_hello() {
-            connection.send_control_frame(ConnectionControl::codec_error(
-                "Hello/Welcome are not allowed after handshake",
+            tx.unbounded_drain(DrainReason::CodecError(
+                "Hello/Welcome are not allowed after handshake".to_owned(),
             ));
             break;
         };
 
         // if it's a control signal, handle it, otherwise, route with message router.
-        if let message::Body::ConnectionControl(ctrl_msg) = &body {
+        if let Body::ConnectionControl(ctrl_msg) = &body {
             // do something
             info!(
                 "Terminating connection based on signal from {}: {}",
                 connection.peer(),
                 ctrl_msg.message
             );
-            if ctrl_msg.signal() == message::Signal::Shutdown {
+            if ctrl_msg.signal() == Signal::Shutdown {
+                tx.unbounded_drain(DrainReason::ConnectionDrain);
                 is_peer_shutting_down = true;
+                needs_drain = true;
             }
```

Comment:
Should we also set `needs_drain = true` if we receive the `Signal::DrainConnection` from the remote peer?

Reply:
Also fixed in future work